### PR TITLE
Introduce per entry stats

### PR DESCRIPTION
**Pull request by ahmetmircik**
ee counterpart is https://github.com/hazelcast/hazelcast-enterprise/pull/3933

### CODE REVIEW

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed.
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   TestFullApplicationContext.testMapConfig:330 expected:<25> but was:<26>
[INFO] 
[ERROR] Tests run: 193, Failures: 1, Errors: 0, Skipped: 0
[INFO] 

[ERROR] There are test failures.
</pre>
</details>

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed.
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.1:checkstyle (default) on project hazelcast: An error has occurred in Checkstyle report generation. Failed during checkstyle execution: There are 2 errors reported by Checkstyle 8.38 with /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/checkstyle/checkstyle.xml ruleset. -> [Help 1]
--------------------------
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/map/impl/record/DataRecordFactory.java:38:5: Cyclomatic Complexity is 14 (max allowed is 12). [CyclomaticComplexity]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/map/impl/record/DataRecordFactory.java:38:5: Return count is 10 (max allowed for non-void methods/lambdas is 8). [ReturnCount]
--------------------------
</pre>
</details>

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed.
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   BasicMapTest.testEntryView:840->HazelcastTestSupport.assertEqualsStringFormat:1375 Expected entryView1.getHits() to be 2, but were -1 expected:<2> but was:<-1>
[ERROR]   EvictionTest.testContainsKeyShouldDelayEviction:844 Diff seconds 0, wait seconds 2
[ERROR]   EvictionTest.testMaxIdle_readThroughBitmapIndex:220->testMaxIdle_readThroughIndex:245
[ERROR]   EvictionTest.testMaxIdle_readThroughOrderedIndex:207->testMaxIdle_readThroughIndex:245
[ERROR]   EvictionTest.testMaxIdle_readThroughUnorderedIndex:214->testMaxIdle_readThroughIndex:245
[ERROR]   ExpirationTimeTest.last_access_time_updated_on_primary_when_read_backup_data_enabled:449 lastAccessTimeBefore:-1, lastAccessTimeAfter:-1
[ERROR]   ExpirationTimeTest.testExpirationTime_whenMaxIdleTime_isBiggerThan_TTL:365 expected:<4999> but was:<1611650658000>
[ERROR]   ExpirationTimeTest.testExpirationTime_whenMaxIdleTime_isSmallerThan_TTL:351 expected:<9998> but was:<1611650666000>
[ERROR]   ExpirationTimeTest.testExpirationTime_withMaxIdleTime:275 expected:<9999> but was:<1611650665000>
[ERROR]   ExpirationTimeTest.testExpirationTime_withMaxIdleTime_afterMultipleAccesses:336 expected:<9999> but was:<1611650659000>
[ERROR]   ExpirationTimeTest.testExpirationTime_withMaxIdleTime_withEntryCustomMaxIdle:289 expected:<9999> but was:<1611650666000>
[ERROR]   ExpirationTimeTest.testExpirationTime_withMaxIdleTime_withEntryCustomMaxIdleGreaterThanConfig:303 expected:<119999> but was:<1611650780000>
[ERROR]   ExpirationTimeTest.testExpirationTime_withNegativeMaxIdleTime:318 expected:<9999> but was:<1611650668000>
[ERROR]   ExpirationTimeTest.testExpirationTime_withNegativeTTL_andMapConfigTTL:242 expected:<59999> but was:<1611650706000>
[ERROR]   ExpirationTimeTest.testExpirationTime_withShorterTTL_andMaxIdle:188 expected:<9999> but was:<1611650672000>
[ERROR]   ExpirationTimeTest.testExpirationTime_withTTL:130 expected:<59999> but was:<1611650721000>
[ERROR]   ExpirationTimeTest.testExpirationTime_withTTL_afterMultipleUpdates:261 expected:<59999> but was:<1611650707000>
[ERROR]   ExpirationTimeTest.testExpirationTime_withTTL_andMapConfigTTL:220 expected:<59999> but was:<1611650710000>
[ERROR]   ExpirationTimeTest.testExpirationTime_withTTL_withShorterMaxIdle:175 expected:<9999> but was:<1611650661000>
[ERROR]   ExpirationTimeTest.testLastAccessTime_isZero_afterFirstPut:376 expected:<0> but was:<-1>
[ERROR]   LocalMapStatsTest_WhenMigration.testHitsGenerated_newNode:62->HazelcastTestSupport.assertTrueEventually:1350->HazelcastTestSupport.assertTrueEventually:1248 expected:<100> but was:<-10>
[ERROR]   DataRecordFactoryTest>AbstractRecordFactoryTest.testNewRecord_withStatisticsEnabledAndCacheDeserializedValuesIsALWAYS:86->HazelcastTestSupport.assertInstanceOf:916 CachedSimpleRecord{cachedValue=null, SimpleRecord{value=HeapData{type=-100, hashCode=-1143465106, partitionHash=-1143465106, totalSize=101, dataSize=93, heapCost=133}, version=0}}  is not an instanceof com.hazelcast.map.impl.record.CachedDataRecordWithStats
[ERROR]   DataRecordFactoryTest>AbstractRecordFactoryTest.testNewRecord_withStatisticsEnabledAndCacheDeserializedValuesIsNEVER:94->HazelcastTestSupport.assertInstanceOf:916 SimpleRecord{value=HeapData{type=-100, hashCode=-1143465106, partitionHash=-1143465106, totalSize=101, dataSize=93, heapCost=133}, version=0} is not an instanceof com.hazelcast.map.impl.record.DataRecordWithStats
[ERROR]   ObjectRecordFactoryTest>AbstractRecordFactoryTest.testNewRecord_withStatisticsEnabledAndCacheDeserializedValuesIsALWAYS:86->HazelcastTestSupport.assertInstanceOf:916 SimpleRecord{value=com.hazelcast.map.impl.record.Person@3c6a660, version=0} is not an instanceof com.hazelcast.map.impl.record.ObjectRecordWithStats
[ERROR]   ObjectRecordFactoryTest>AbstractRecordFactoryTest.testNewRecord_withStatisticsEnabledAndCacheDeserializedValuesIsNEVER:94->HazelcastTestSupport.assertInstanceOf:916 SimpleRecord{value=com.hazelcast.map.impl.record.Person@3c6a660, version=0} is not an instanceof com.hazelcast.map.impl.record.ObjectRecordWithStats
[ERROR]   LazyEvictableEntryViewTest.test_getCreationTime:97 expected:<0> but was:<-1>
[ERROR]   LazyEvictableEntryViewTest.test_getHits:107 expected:<0> but was:<-1>
[ERROR]   LazyEvictableEntryViewTest.test_getLastAccessTime:112 expected:<0> but was:<-1>
[ERROR]   LazyEvictableEntryViewTest.test_getLastStoredTime:117 expected:<0> but was:<-1>
[ERROR]   LazyEvictableEntryViewTest.test_getLastUpdateTime:122 expected:<0> but was:<-1>
[ERROR]   MapSplitBrainTest>SplitBrainTestSupport.testSplitBrain:195->SplitBrainTestSupport.doIteration:207->onAfterSplitBrainHealed:185->afterMergeHigherHitsMergePolicy:249 expected:<[higherHitsV]alue1> but was:<[v]alue1>
[ERROR]   MapSplitBrainTest>SplitBrainTestSupport.testSplitBrain:195->SplitBrainTestSupport.doIteration:207->onAfterSplitBrainHealed:187->afterMergeLatestAccessMergePolicy:291 expected:<[LatestAccessedV]alue2> but was:<[v]alue2>
[ERROR]   MapSplitBrainTest>SplitBrainTestSupport.testSplitBrain:195->SplitBrainTestSupport.doIteration:207->onAfterSplitBrainHealed:189->afterMergeLatestUpdateMergePolicy:320 expected:<[LatestUpdatedV]alue2> but was:<[v]alue2>
[INFO] 
[ERROR] Tests run: 35547, Failures: 33, Errors: 0, Skipped: 986
[INFO] 

[ERROR] There are test failures.
</pre>
</details>

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed.
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.1:checkstyle (default) on project hazelcast: An error has occurred in Checkstyle report generation. Failed during checkstyle execution: There is 1 error reported by Checkstyle 8.38 with /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/checkstyle/checkstyle.xml ruleset. -> [Help 1]
--------------------------
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/test/java/com/hazelcast/map/LocalMapStatsTest_WhenMigration.java:73:10: ',' is not followed by whitespace. [WhitespaceAfter]
--------------------------
</pre>
</details>

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed.
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   BasicMapTest.testEntryView:840->HazelcastTestSupport.assertEqualsStringFormat:1375 Expected entryView1.getHits() to be 2, but were -1 expected:<2> but was:<-1>
[INFO] 
[ERROR] Tests run: 35547, Failures: 1, Errors: 0, Skipped: 986
[INFO] 

[ERROR] There are test failures.
</pre>
</details>

**Comment by ahmetmircik** [comment]
Checkstyle errors will automatically be fixed after merging https://github.com/hazelcast/hazelcast/pull/18099

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed.
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   EntryCostEstimatorTest.testExactHeapCostAfterUpdateWithMultipleBackupNodes:80 Heap cost calculation is wrong! expected:<152> but was:<128>
[ERROR]   EntryCostEstimatorTest.testSinglePut:67 expected:<152> but was:<128>
[INFO] 
[ERROR] Tests run: 35658, Failures: 2, Errors: 0, Skipped: 984
[INFO] 

[ERROR] There are test failures.
</pre>
</details>

File: hazelcast/src/test/java/com/hazelcast/map/BasicMapTest.java [hunk 1]
```diff
@@ -131,10 +131,19 @@ public void init() {
 
     protected Config getConfig() {
         Config cfg = smallInstanceConfig();
-        cfg.getMapConfig("default").setStatisticsEnabled(statisticsEnabled);
+        cfg.getMapConfig("default")
+                .setStatisticsEnabled(statisticsEnabled)
+                .setPerEntryStatsEnabled(statisticsEnabled);
```
Comment (blazember) [inline hazelcast/src/test/java/com/hazelcast/map/BasicMapTest.java:136]: Maybe add some comment to the field that we intentionally use it for controlling both per-map and per-entry statistics. With this approach, we test both the presence/absence of the two stats. 

Btw, do we have any test for the `per-map:off`, `per-entry:on` configuration?

File: hazelcast/src/test/java/com/hazelcast/map/LocalMapStatsTest_WhenMigration.java [hunk 2]
```diff
@@ -67,7 +70,7 @@ public void run() throws Exception {
 
                 assertEquals(100, hits1 + hits2);
             }
-        });
+        }, 20);
```
Comment (blazember) [inline hazelcast/src/test/java/com/hazelcast/map/LocalMapStatsTest_WhenMigration.java:73]: Leftover?

File: hazelcast/src/test/java/com/hazelcast/map/impl/record/AbstractRecordFactoryTest.java [hunk 3]
```diff
@@ -118,7 +118,7 @@ InternalSerializationService createSerializationService() {
     protected MapContainer createMapContainer(boolean isStatisticsEnabled,
                                             CacheDeserializedValues cacheDeserializedValues) {
         MapConfig mapConfig = new MapConfig()
-                .setStatisticsEnabled(isStatisticsEnabled)
+                .setPerEntryStatsEnabled(isStatisticsEnabled)
```
Comment (blazember) [inline hazelcast/src/test/java/com/hazelcast/map/impl/record/AbstractRecordFactoryTest.java:121]: Consider renaming the argument.

**Review by blazember — COMMENTED** [review]
Went through the PR ignoring the changes from the parent PR.

In general, the change looks good, added only some minors. Please add the change to `full-example` XML and YAML configs as well.

Please ping me if the parent PR is merged and I take a fresh look.

File: hazelcast/src/test/java/com/hazelcast/map/LocalMapStatsTest_WhenMigration.java (hunk 2 above)
Comment (ahmetmircik) [reply hazelcast/src/test/java/com/hazelcast/map/LocalMapStatsTest_WhenMigration.java]: deleted.

File: hazelcast/src/test/java/com/hazelcast/map/impl/record/AbstractRecordFactoryTest.java (hunk 3 above)
Comment (ahmetmircik) [reply hazelcast/src/test/java/com/hazelcast/map/impl/record/AbstractRecordFactoryTest.java]: done.

File: hazelcast/src/test/java/com/hazelcast/map/BasicMapTest.java (hunk 1 above)
Comment (ahmetmircik) [reply hazelcast/src/test/java/com/hazelcast/map/BasicMapTest.java]: Updated test parameter list to reflect all combinations

File: hazelcast/src/main/java/com/hazelcast/map/impl/record/ObjectRecordFactory.java [hunk 4]
```diff
@@ -39,7 +39,7 @@ public ObjectRecordFactory(MapContainer mapContainer,
     @Override
     public Record<Object> newRecord(Object value) {
         MapConfig mapConfig = mapContainer.getMapConfig();
-        boolean statisticsEnabled = mapConfig.isStatisticsEnabled();
+        boolean statisticsEnabled = mapConfig.isPerEntryStatsEnabled();
```
[the remark this answers is not in the document]
Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/map/impl/record/ObjectRecordFactory.java]: fixed.

File: hazelcast/src/main/java/com/hazelcast/map/impl/record/DataRecordFactory.java [hunk 5]
```diff
@@ -38,7 +38,7 @@ public DataRecordFactory(MapContainer mapContainer, SerializationService ss) {
     @Override
     public Record<Data> newRecord(Object value) {
         MapConfig mapConfig = mapContainer.getMapConfig();
-        boolean statisticsEnabled = mapConfig.isStatisticsEnabled();
+        boolean statisticsEnabled = mapConfig.isPerEntryStatsEnabled();
```
[the remark this answers is not in the document]
Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/map/impl/record/DataRecordFactory.java]: fixed

File: docs/design/map/01-per-entry-stats.md [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+Entry-level stats are: `hits-lastAccessTime-lastUpdateTime-creationTime-lastStoredTime`
+ 
+### Replication of Map Level Stats
+Map-level stats are created per map's partition. They are
+replicated to new nodes to be safe from partition ownership changes.
+
+Entry-level stats are already replicated
+when enabled. No change is needed for them.
+
+### Metrics
```
[the remark this answers is not in the document]
Comment (ahmetmircik) [reply docs/design/map/01-per-entry-stats.md]: I don't have any reason to add it to metrics. Removed that section for now.

**Review by petrpleshachkov — APPROVED** [review]

File: hazelcast/src/main/java/com/hazelcast/config/MapConfig.java [hunk 7]
```diff
@@ -912,5 +971,9 @@ public void readData(ObjectDataInput in) throws IOException {
         merkleTreeConfig = in.readObject();
         eventJournalConfig = in.readObject();
         metadataPolicy = MetadataPolicy.getById(in.readShort());
+
+        if (in.getVersion().isGreaterOrEqual(Versions.V4_2)) {
+            perEntryStatsEnabled = in.readBoolean();
+        }
```
[the remark this answers is not in the document]
Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/config/MapConfig.java:977]: In that case, if server is v4.2 and cluster is v4.1, we create records which have stats on it regardless of the config. So no issue i expect.

File: docs/design/map/01-per-entry-stats.md [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+Based on this, now we have a chance to enable/disable
+record metadata to reduce memory usage of a record.
+
+## Design
+
+We categorize stats in two types: map-level and entry-level.
+Map level is enabled by default but entry-level is not.
+
+Map-level stats are represented with class `LocalRecordStoreStatsImpl`
+It has 3 fields: `hits-lastAccess-lastUpdate`
```
[the remark this answers is not in the document]
Comment (ahmetmircik) [reply docs/design/map/01-per-entry-stats.md:21]: done.

File: docs/design/map/01-per-entry-stats.md [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+## Design
+
+We categorize stats in two types: **map-level** and **entry-level**.
+Map level is enabled by default but entry-level is not.
+
+Map-level stats are:
+- **hits:** total hits of map
+- **lastAccessTime:** last access time to map
+- l**astUpdateTime:** last update time of map
```
Comment (sancar) [inline docs/design/map/01-per-entry-stats.md:23]: There is a typo here
l**astUpdateTime: -> **lastUpdateTime:

**Review by sancar — APPROVED** [review]

**Review by sancar — CHANGES_REQUESTED** [review]
There is a problem on the protocol side that needs to be fixed. 
https://github.com/hazelcast/hazelcast-client-protocol/pull/372

File: docs/design/map/01-per-entry-stats.md (hunk 9 above)
Comment (ahmetmircik) [reply docs/design/map/01-per-entry-stats.md]: fixed

**Review by sancar — APPROVED** [review]

**Review by blazember — APPROVED** [review]

**Pull request merged**